### DISCUSSION TITLE
vba-articles/performance-tips

### DISCUSSION THREAD

**Discussion by giscus** [sancarn/vba-articles#4] [_giscus]
# vba-articles/performance-tips

How to improve performance of VBA macros is a common question on the VBA reddit. In this section we will provide examples of code, and test their performance against eachother. We believe it is best to work by example and hope this will give you ideas of how you can improve performance in your projects.

https://sancarn.github.io/vba-articles/performance-tips.html

**Comment by KallunWillock** [comment]
I don't remember seeing the Advanced Array Manipulation section the first time I read this. Very helpful.

**Comment by Excel-lent** [comment]
I'm currently reading the "Performance Tips for VBA" and have few comments.
I'm sure the article is still in progress and being improved, but I have few ideas on how to improve it.

- It would be nice to hide the code and performance measurements in collapsible sections. This would make the article significantly shorter and easier to read.

- It would be better to measure the performance in percentages and display them in descending order in the bar chart.

I also had a few additional performance tips, but at the moment (at home) I can't reproduce them. For example, using “Cells(1, 1).Value2 = x” except “Cells(1, 1) = x” or using “Long” except “Integer”. I'm pretty sure they depend on the Excel version (at work I have Excel 2019, at home - Excel 2007). Therefore, it would also be great to collect statistics from different versions of Excel. Additionally, it would also be great to have an Excel spreadsheet with all tests somewhere under “Releases” so that instead of adding all the modules and classes, users can open it and run the script that automatically measures performance and sort them in descending order.

**Reply by KallunWillock** [reply]
I like these suggestions - you can never go wrong with a bar chart!

**Reply by sancarn** [reply]
Feel free to [contribute](https://github.com/sancarn/vba-articles/blob/main/_posts/2023-09-01-performance-tips.md)! I agree with most of your points.

> hide the code and performance measurements in collapsible sections

I agree, this would be beneficial. Need to investigate how easy this would be to do with jekyll.

> It would be better to measure the performance in percentages

Percentages of what? Of the slowest?

> Visualise with a bar chart

Would be cool, again would need to investigate how easy this is to do with jekyll. I imagine we could create a custom element to make this easier.

> it would also be great to collect statistics from different versions of Excel

I agree, but that might be quite hard as it also depends on the user's system... But perhaps if you calculate all percentages based on the same benchmark, then that will allow us to scale the performance correctly across multiple different measures?

> have an Excel spreadsheet with all tests, and allow users to run it

100% agree with this! Though don't think it should be in releases, perhaps a release in it's own seperate repo :) If stored on this repo, it should be in ./assets/bin

**Comment by KallunWillock** [comment]
As regards collapsible sections, this would work, no? https://www.endtoend.ai/tutorial/collapsible-code-blocks/

**Reply by Excel-lent** [reply]
Yes, it should work. At least [here](https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab) it works fine.

**Comment by Excel-lent** [comment]
Hello Sancarn and Kallun,

I reworked a little bit the description of performance tips and added a table that everyone can use as reference with his / her own Excel version (see https://github.com/Excel-lent/VBA-performance-tests).
Since the changes in pull request will be quite huge I don’t feel myself “free” to update your repository and need your agreement 😎
Another reason to put the description separately is because I added the comments inside (marked in red and blue for important and not very important comments).
The pictures are unfortunately still in old PNG format because my nice SVG library is not completely ready.

Have nice day and Happy New Year! 🥳

**Reply by sancarn** [reply]
Hi @Excel-lent, 

Thanks for putting in all the effort to improve the article :) Happy new year!

I love the spoilers for the code samples. Tables with different versions is a nice addition. Like the refreshable excel file. 

I do think SA1 and SA2 should remain, mainly because I don't think people would think it obvious that `CreateObject()` is euquivalent speed to `new Class`. That said these should likely be inside their own section of "things that don't affect performance" perhaps? Or just a seperate article all together...

> The pictures are unfortunately still in old PNG format because my nice SVG library is not completely ready.

Which pictures?

### About percentages

I'm not 100% convinced on the percentages - If I see 200% what does that mean? 2x slower? 

| Sancarn | 2007/32 | 2010/32 | 2019/64|
|----------|-----------|----------|----------|
| 294%     | 54%        | 86%      | 192%     |

Like I'm not 100% sure what we are saying here... Are we saying that in 2007/32 bit using `cell.select: selection.value = "hello"` was faster than `cell.value = "hello"`?! If so that sounds kinda mental/difficult to believe...

In my mind I feel that users would be confused if 200% is better or worse than 50%... At least with time it's clear that lower is better. But I do get your point that time is all relative to the system it's run on...

### Workbook

Workbook is a little messy, but I do like the graphs especially, sorted to show what was most effective and what wasn't. Not sure what the tabs "Performance Sancarn" etc. is about but imagine this is because these need to be converted to perfect basically?

----

Feel free to submit a PR as it'd allow me to work on it and make any changes i feel necessary.

**Reply by KallunWillock** [reply]
> The pictures are unfortunately still in old PNG format because my nice SVG library is not completely ready.
> 
> Have nice day and Happy New Year! 🥳

Happy New Year to you too! I'm looking forward to your SVG library, btw. I've also been working on something SVG related, but more in terms of rendering it. Haven't quite wrapped my head around SVG code yet, but it's a work in progress.

**Reply by Excel-lent** [reply]
Sorry for some confusion. At the beginning I wanted to add performance diagram as a picture, but now I have doubts.
I uploaded additional picture to section `ST` today.

- Calculation of performance gain in percents:
$$\frac{t_{slow}}{t_{fast}}-1$$

For example, if we have 100 ms execution time in case of "slow" function and 50 ms in case of "fast" function, 

$$\frac{100 \ ms}{50 \ ms}-1 = 1,$$
so the performance gain is 100%.

I think the formula is correct for situations when the "slow" function is really slower than the "fast" function. In opposite direction ("slow" is faster than "fast") the smallest value is -100%. I think it is wrong and I have to use in this case another formula:
$$1-\frac{t_{fast}}{t_{slow}}$$

You are right, the descriptions in the table and in the document are misleading. If I will change "Direct Value Setting" to "Direct value setting vs select and set", the value of 294% of performance increasing is self-explaining.

- The sheets "Performance Sancarn" and "Measurements Sancarn" are copies of "Performance" and "Measurements" with execution times from original article. In a new version I deleted them. 
BTW, which Excel version do you use?

- I added deleted parts to S6 (`CreateObject()` vs `new Class`).

**Reply by TarVK** [reply]
I feel that with the codeblocks fully hidden, it might be more effort to scan the document. Maybe the code blocks could remain hidden, but a small preview codeblock with the essence of the topic could be added, for instance for the cut/paste topic, one could add:
```vba
cells(1, 1).cut
cells(1, 2).select
ActiveSheet.paste
```

This way unnecessary details are still hidden by default, while keeping the benefit of having tangible code visible rather than just titles and discriptions

**Reply by sancarn** [reply]
> BTW, which Excel version do you use?

Excel for O365 | 64 Bit

> a small preview codeblock with the essence of the topic

@TarVK That's fair, and tbf we are there for the code in the first place... 👀 Unless we can fully understand at a glance what we are comparing from the title at least some code block should be visible...

**Reply by Excel-lent** [reply]
@TarVK, yes, this has more sense than just to hide the code blocks.
What do you think if the important parts of the code will be shown as a small table? Like this:
> |       Select and set      |         Set directly        |
> |---------------------------|-----------------------------|
> | cells(1, 1).select        | cells(1, 1).value = "hello" |
> | selection.value = "hello" |                             |

Then the users will immediately see how to speedup the code.

**Reply by sancarn** [reply]
I think we'd be better off using Jekyll or here if possible, to show 2 code blocks side-by-side rather than a markdown table... 

To be fair though at that point we might just be better off generating the site's html or something...

**Reply by Excel-lent** [reply]
Yes, this would be a good opportunity to automatically generate markdowns from VBA output.
On the other hand, the values in the document are qualitative, not quantitative, and every time a small change is made in the table, the document is regenerated.
I prefer to create the document once and only edit changed values/code.
According to @TarVK comment, my suggestion is to add blocks like this:
<table>
<tr>
<th>Select and set</th>
<th>Set directly</th>
</tr>
<tr>
<td>
  
```vb
cells(1, 1).select
selection.value = "hello"      
```

</td>
<td>

```vb
cells(1, 1).value = "hello"      


```

</td>
</tr>
</table>

**Reply by sancarn** [reply]
Yeah, that works for me! So we have the comparrison blocks, but also the hidden full source code beneath?

**Reply by Excel-lent** [reply]
I think yes. We need hidden blocks to show how the performance measurements were done.
I would not like to search the whole source to see how these few lines are implemented for measurements.

**Comment by Excel-lent** [comment]
The colors are visible only in VScode.

**Comment by Excel-lent** [comment]
Hi everybody!
Sorry for the delay, but I hope, I have implemented all what I wanted to implement. 
I have made a PR, but there are still two external links: one is to a picture in “String operations”. Another one is the reference to the table (I was not sure where to place it).
Also, I’ve found that to sustain that big project was not easy. That's why I split the MD file into several smaller MDs. After change of one MD file you have to run compile.bat to create a new README.MD. The complete set you can download from https://github.com/Excel-lent/VBA-performance-tests/releases/download/v0.0.5-beta/MD.s.zip.
When you will resolve these dependencies, I will remove my temporary repository.

**Reply by sancarn** [reply]
Hi @Excel-lent, thanks so much for getting back to this project! And thanks for the PR - makes my life a little easier 😄 

Definitely need to make it easier to maintain the sections. Splitting them into MD files to compile is a good idea, either that or shapes/excel tabs might be a good way forward.

I'm still not 100% sold on the percentages... And stuff like this seems super confusing to me:

> | 2007/32 | 2010/32 | 2019/64 |  365/64 |
> |:-------:|:-------:|:-------:|:-------:|
> |   196%  |   55%   |   269%  |   305%  |

I'm wondering if it'd still be nice to have the times in there:

> | Key | 2007/32 | 2010/32 | 2019/64 |  365/64 |
> |-----|:-------:|:-------:|:-------:|:-------:|
> |  `#1` (ms) |   10  |   ...   |   ...  |   ... |
> |  `#2` (ms) |   5.1  |   ....   |   ...  |   ... |
> |   % |   196%  |   55%   |   269%  |   305%  |

On the whole though it looks good at the moment so I think I will apply the PR and maybe think about how we are going to maintain this and update it later.

**Reply by sancarn** [reply]
Oh poop, looks like something is up with the code blocks in tables for this jekyll layout/style >.<

![image](https://github.com/sancarn/vba-articles/assets/7938900/e081b702-5518-4d8e-8cf2-069d3a73914c)
